Feature/liked

leejaeeun59357 opened the pull request:
### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
**AS-IS**
- 회원가입
- 로그인
- 게시글
  - 작성 기능
  - 수정 기능
  - 삭제 기능
  - 조회 기능
- 댓글
  - 작성 기능
  - 수정 기능
  - 삭제 기능

**TO-BE**
- 게시글, 댓글에 좋아요를 할 수 있는 기능 구현
  - 로그인한 사용자만 좋아요 기능 사용 가능
  - 이미 좋아요를 누른 게시글이나 댓글에 대해 다시 한 번 좋아요를 하게 되면 취소되도록 설정함

### 테스트
- [x] 로그인된 사용자만 좋아요 가능
- [x] 이미 좋아요된 게시글, 댓글 다시 한 번 좋아요 시 삭제됨

Commit:
[feat] 게시글, 댓글의 좋아요 기능 구현

- 게시글과 댓글에 좋아요를 누를 수 있는 기능 구현
- 이미 좋아요가 되어 있는 게시글, 댓글에 대해 다시 한 번 좋아요를 하게 되면 취소되도록 설정함

Commit:
Update README.md

mingming-mentor commented on src/main/java/org/board/exercise_board/Liked/domain/repository/LikedRepository.java:
```diff
@@ -0,0 +1,11 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface LikedRepository extends JpaRepository<Liked, Long> {
+  Boolean existsByTypeAndTypeIdAndUser(Type type, Long typeId, User user);
```
exist면 true/false만 나올 것 같아요. 그러면 굳이 Boolean으로 boxing하지 않고 boolean을 사용해도 좋을 것 같습니다

leejaeeun59357 replied:
예! 수정하겠습니다!

mingming-mentor commented on src/main/java/org/board/exercise_board/Liked/service/LikedService.java:
```diff
@@ -0,0 +1,55 @@
+  public String saveLiked(Type type, Long id, String loginId) {
+
+    // 게시글이 존재하지 않을 때
+    if(type == Type.POST) {
+      var result = postService.findPost(id);
+
+    } else if(type == Type.COMMENT) {
+
+      // 댓글이 존재하지 않을 때
+      var result = commentService.findComment(id);
+    }
```
이 부분을 adapter pattern을 적용해도 좋을 것 같아요. https://jusungpark.tistory.com/22

mingming-mentor commented on src/main/java/org/board/exercise_board/Liked/service/LikedService.java:
```diff
@@ -0,0 +1,55 @@
+  private final CommentService commentService;
+  private final UserService userService;
+
+  public String saveLiked(Type type, Long id, String loginId) {
```
좋아요를 빠르게 2번 눌렀을 경우, `좋아요` -> `취소` 해서 취소가 된 상태가 되어야할 것 같은데
다중 쓰레드에서는 like로 save 될수도 있을 것 같아요. 
첫번째 쓰레드가 `likedRepository.existsByTypeAndTypeIdAndUser(type,id,user)` ~ `save` 사이에 두번째 쓰레드가 `likedRepository.existsByTypeAndTypeIdAndUser(type,id,user)`를 호출할 경우에요.
좀더 정교하게 개발하려면 lock을 거시면 좋을 것 같습니다

leejaeeun59357 replied:
예! 수정하겠습니다!

mingming-mentor reviewed (approved):
수고 많으셨습니다!

Commit:
[refactor] 메서드 lock 걸기 및 반환값 변경

- 다중 스레드 사용 시 발생할 수 있는 문제점을 예방하기 위해 synchronized 블럭을 사용하여 lock 걸기
- boxing 된 Boolean이 아닌 boolean으로 반환값 변경

Commit:
[refactor] 메서드 이름 변경

- FindSomething 인터페이스를 상속받도록 PostService와 CommentService 변경
- PostService 의 메서드와 CommentService의 메서드 이름 변경

Commit:
[feat] findSomething 인터페이스 생성

Commit:
[refactor] 클린코드로 수정

Commit:
[feat] 어댑터 패턴 사용하여 코드 구조 변경
